### ISSUE TITLE
Upgrade to using Python 3.10.4

### ISSUE THREAD

**Issue by lisagoodrich** [dtcenter/METcalcpy#270]
## Describe the Task ##
Update any code necessary for upgrading to Python 3.10.4.  We are taking this action because WCOSS2 is only allowing Python 3.10.4 to be used for operational installations.  Therefore, we need to ensure that our next release uses Python 3.10.4.

**Please also update .readthedocs.yaml with 3.10.4**

### Time Estimate ###
*Estimate the amount of work required here.*
*Issues should represent approximately 1 to 3 days of work.*
**(Assignee please fill this in once you know.)**

### Sub-Issues ###
Consider breaking the task down into sub-issues.
- [ ] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
Incorporate in a beta version as soon as possible to ensure there are no problems.

### Funding Source ###
Please ask Tara before starting on this work.

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED PROJECT ASSIGNMENT** label
- [x] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdataio](https://github.com/dtcenter/METdataio/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)

## Task Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Add any new Python packages to the [METplus Components Python Requirements](https://metplus.readthedocs.io/en/develop/Users_Guide/overview.html#metplus-components-python-requirements) table.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issues
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

**Comment by bikegeek** [comment]
numerous tests are failing because the Iterable abstract class was removed from collections in Python 3.10 
Error messages from pytest output:

 

> ImportError: cannot import name 'Iterable' from 'collections'

(https://stackoverflow.com/questions/72032032/importerror-cannot-import-name-iterable-from-collections-in-python):

> Remove deprecated aliases to [Collections Abstract Base Classes](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes) from the collections module. (Contributed by Victor Stinner in bpo-37324.)


**https://docs.python.org/3.10/library/collections.abc.html#collections.abc.Iterable**
class collections.abc.Iterable

    ABC for classes that provide the __iter__() method.

    Checking isinstance(obj, Iterable) detects classes that are registered as [Iterable](https://docs.python.org/3.10/library/collections.abc.html#collections.abc.Iterable) or that have an __iter__() method, but it does not detect classes that iterate with the __getitem__() method. The only reliable way to determine whether an object is [iterable](https://docs.python.org/3.10/glossary.html#term-iterable) is to call iter(obj).


Fix from StackOverflow:

> You can use Iterable from collections.abc instead, or use Python 3.9 if the problem is in a dependency that can't be updated.

**Comment by bikegeek** [comment]
updates (to address Iterable issue) in the following:
agg_eclv.py
agg_stat.py
bootstrap.py

**Comment by bikegeek** [comment]
Additional issues with the test_spacetime.py tests:
ImportError: /d1/personal/mwin/miniconda3/envs/metcalcpy_py310/lib/python3.10/site-packages/numexpr/../../../libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /d1/personal/mwin/miniconda3/envs/metcalcpy_py310/lib/python3.10/site-packages/scipy/fft/_pocketfft/pypocketfft.cpython-310-x86_64-linux-gnu.so)

From top of traceback:
ImportError while importing test module '/d1/personal/mwin/METdataio_Tutorial/METcalcpy/test/test_spacetime.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
../../../miniconda3/envs/metcalcpy_py310/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
test_spacetime.py:12: in <module>

**Comment by bikegeek** [comment]
Additional issues with the test_spacetime.py tests:
ImportError: /d1/personal/mwin/miniconda3/envs/metcalcpy_py310/lib/python3.10/site-packages/numexpr/../../../libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /d1/personal/mwin/miniconda3/envs/metcalcpy_py310/lib/python3.10/site-packages/scipy/fft/_pocketfft/pypocketfft.cpython-310-x86_64-linux-gnu.so)

From top of traceback:
ImportError while importing test module '/d1/personal/mwin/METdataio_Tutorial/METcalcpy/test/test_spacetime.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
../../../miniconda3/envs/metcalcpy_py310/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
test_spacetime.py:12: in <module>

**Comment by bikegeek** [comment]
Upon further inspection of the agg_eclv.py code, the import of bootstrap is where the issue lies.  Replace this import:
_from collections import Iterable_

with 

_from collections.abc import Iterable_

This also corrects the issue with the agg_stat.py import error.

**Comment by bikegeek** [comment]
Cannot update .readthedocs.yaml to python 3.10.  The highest supported version of Python in ReadtheDocs is Python 3.8

**Comment by jprestop** [comment]
> Cannot update .readthedocs.yaml to python 3.10. The highest supported version of Python in ReadtheDocs is Python 3.8

Thank you, @bikegeek. I was thinking about creating a separate issue for this for the final beta release to see if it is something we can update then.  What do you think about that?

**Comment by bikegeek** [comment]
Performance diagram test is failing because in performance_diagram.py, when the linewidth is set to 'NA' in the config file (to turn off the connecting lines), the linewidth value is set to 0.  Newer versions of matplotlib do NOT allow this.  Instead, set linewidth = None.
